### PR TITLE
Refactor recommendation code to include generate Recommendation API

### PR DESCRIPTION
**Pull request by khansaad**
This PR refactors `generate Recommendations` API code.

This is a copy of #1161 raised against `mvp_demo`

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+            sendErrorResponse(response, e, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage());
+        } finally {
+            if (timerBUpdateRecommendations != null) {
+                MetricsConfig.timerUpdateRecomendations = MetricsConfig.timerBUpdateRecommendations.tag(KruizeConstants.DataSourceConstants.DataSourceQueryJSONKeys.STATUS, statusValue).register(MetricsConfig.meterRegistry());
+                timerBUpdateRecommendations.stop(MetricsConfig.timerUpdateRecomendations);
+            }
+        }
+    }
+
+    public static void fetchMetricsBasedOnDatasource(KruizeObject kruizeObject, Timestamp interval_end_time, Timestamp interval_start_time, DataSourceInfo dataSourceInfo) throws Exception {
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java:132]: Add a javadoc and move this to RecEngine. Also add a TODO as part of the javadoc to breakup this code into the right abstractions

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java]: Done

**Comment by dinogun** [comment]
@khansaad Please add documentation for the `generateRecommendation` API

File: src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
+/*******************************************************************************
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java:1]: This needs to be added to the servlets in [services](https://github.com/kruize/autotune/blob/master/src/main/java/com/autotune/analyzer/Analyzer.java).

Comment (khansaad) [reply src/main/java/com/autotune/analyzer/services/GenerateRecommendations.java:1]: Added now

**Review by dinogun — APPROVED** [review]
LGTM

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 3]
```diff
@@ -199,43 +219,37 @@ public KruizeObject prepareRecommendations(int calCount) {
...
-                    kruizeObject.setValidation_data(validationOutputData);
-                } catch (Exception e) {
+            // set the default terms if the terms aren't provided by the user
+            if (kruizeObject.getTerms() == null)
+                KruizeObject.setDefaultTerms(terms, kruizeObject);
+            // set the performance profile
+            setPerformanceProfile(kruizeObject.getPerformanceProfile());
+            // get the datasource
+            // TODO: temporarily setting the only supported datasource, this needs to fetched from the KruizeObject
+            String dataSource = KruizeConstants.SupportedDatasources.PROMETHEUS;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:229]: get datasource from experiment 

String dataSource = kruizeObject.getDataSource();

**Pull request merged**